### ISSUE TITLE
Convert to LXML on server

### ISSUE THREAD

**Issue by gvwilson** [swcarpentry/DEPRECATED-website#212]
1. Merge @gvwilson's change-xml-parser branch.
2. Install lxml and html5lib on software-carpentry.org.
3. Test.
4. Install.
5. Add instructions to README.md so that other people know what they need to install and do.

**Comment by wking** [comment]
See #168 for the root cause of this PR.

**Comment by pipitone** [comment]
So, after setting the PYTHONPATH like so,

``` bash
export PYTHONPATH=~/.python
```

I used [this script](http://gsnedders.com/installing-lxml-on-dreamhost) to install lxml on dreamhost: 

``` bash
#!/bin/sh
mkdir work
cd work
curl -O ftp://xmlsoft.org/libxml2/LATEST_LIBXML2
curl -O ftp://xmlsoft.org/libxml2/LATEST_LIBXSLT
tar -xzf LATEST_LIBXML2
tar -xzf LATEST_LIBXSLT
cd libxml2-*
./configure --prefix=${HOME}/.local
make
make install
cd ../libxslt-*
./configure --prefix=${HOME}/.local --with-libxml-prefix=${HOME}/.local --with-libxml-include-prefix=${HOME}/.local/include --with-libxml-libs-prefix=${HOME}/.local/lib
make
make install
export LDFLAGS=-L/$HOME/.local/lib
export CPPFLAGS=-I/$HOME/.local/include
export LD_RUN_PATH=$HOME/.local/lib
curl -O http://peak.telecommunity.com/dist/ez_setup.py
python ez_setup.py
easy_install --prefix=~/.python lxml
python -c "from lxml import etree"
```

Then I ran, 

``` bash
easy_install --prefix=~/.python html5lib
```

to install html5lib.

Will test shortly.

**Comment by pipitone** [comment]
er, that PTYHONPATH should read: 

``` bash
export PYTHONPATH=~/.python/lib/python2.6/site-packages
```

**Comment by pipitone** [comment]
Okay, when I get the following error running this on dreamhost:

```
[amherst]$ make OUT_DIR=/home/swcarpentry/dev.software-carpentry.org/ SITE=http://dev.software-carpentry.org check
...
3_0/gui.html
3_0/client.html
3_0/server.html
3_0/empirical.html
3_0/lifecycle.html
3_0/summary.html
3_0/license.html
3_0/ack.html
3_0/bib.html
make[1]: Entering directory `/home/swcarpentry/website'
Traceback (most recent call last):
  File "bin/links.py", line 88, in <module>
    main(sys.argv[1])
  File "bin/links.py", line 80, in main
    links = get_links(root_dir, [f for f in filenames if f.endswith('.html')])
  File "bin/links.py", line 54, in get_links
    doc = read_xml(f)
  File "/home/swcarpentry/website/bin/util.py", line 22, in read_xml
    sys.stderr.write('errors in {}\n'.format(filename))
ValueError: zero length field name in format
make[1]: *** [check-links] Error 1
make[1]: Leaving directory `/home/swcarpentry/website'
make: *** [check] Error 2
```

This looks to be a problem with the use of a format string without positional argument specifiers, something that was [only introduced in python 2.7](http://docs.python.org/2/library/string.html#format-string-syntax).  Dreamhost runs python 2.6.6 (we can upgrade, if we want).

This fix: 

``` diff
diff --git a/bin/util.py b/bin/util.py
index 9517fa3..3308ec5 100644
--- a/bin/util.py
+++ b/bin/util.py
@@ -19,9 +19,9 @@ def read_xml(filename, mangle_entities=False):
     doc = html5parser.parse(filename, parser=parser)

     if parser.errors:
-        sys.stderr.write('errors in {}\n'.format(filename))
+        sys.stderr.write('errors in {0}\n'.format(filename))
         for e in parser.errors:
-            sys.stderr.write('    {}\n'.format(e))
+            sys.stderr.write('    {0}\n'.format(e))

     return doc
```

then produces: 

```
...
3_0/empirical.html
3_0/lifecycle.html
3_0/summary.html
3_0/license.html
3_0/ack.html
3_0/bib.html
make[1]: Entering directory `/home/swcarpentry/website'
errors in /home/swcarpentry/dev.software-carpentry.org/blog/2013/01/plos-ad.html
    ((2, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/about/index.html
    ((1, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/bootcamps/faq.html
    ((1, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/blog/2010/12/slides-for-five-five-oo-episodes-online.html
    ((2, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/blog/2013/01/teaching-commercially.html
    ((2, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/blog/2010/09/will-america%e2%80%99s-universities-go-the-way-of-its-car-companies.html
    ((2, 109), 'unknown-doctype', {})
errors in /home/swcarpentry/dev.software-carpentry.org/blog/journal.html
    ((1, 109), 'unknown-doctype', {})
make[1]: Leaving directory `/home/swcarpentry/website'
make[1]: Entering directory `/home/swcarpentry/website'
make[1]: Leaving directory `/home/swcarpentry/website'
make[1]: Entering directory `/home/swcarpentry/website'
python bin/journal.py /home/swcarpentry/dev.software-carpentry.org//blog/*/*/*.html > /home/swcarpentry/dev.software-carpentry.org//blog/journal.html
make[1]: Leaving directory `/home/swcarpentry/website'
```

So it looks like it's working.

**Comment by jiffyclub** [comment]
Sheesh, why is it so hard to install `lxml` on dreamhost? Is there any way we could use `pip` and virtualenv?

(And my bad on the 2.6 incompatible code there. We should upgrade to Python 2.7 if we can.)

**Comment by wking** [comment]
On Mon, Jan 14, 2013 at 09:09:08AM -0800, pipitone wrote:

> This looks to be a problem with the use of a format string without
> positional argument specifiers, something that was only introduced
> in python 2.7.

These format strings are an easy enough fix that they shouldn't be a
reason to break 2.6 support.

On Mon, Jan 14, 2013 at 09:15:06AM -0800, Matt Davis wrote:

> We should upgrade to Python 2.7 if we can.

+1 on upgrading, although my preference would be for something from
Python 3 ;).  Once we get the build toolchain converted to Pelican,
other people will be handling Python-version compatibility for us, so
it shouldn't be as much of a problem.  That said, there are plans to
drop Python 2.6 support from Pelican as part of their move to add
Python 3 support [1](https://github.com/getpelican/pelican/commit/4ac094966e06eac5d540debbd166162297a768d4).

**Comment by gvwilson** [comment]
As far as I can tell, only Python 2.6 and 2.7 are installed on dreamhost 
right now --- and yeah, we can virtualenv if it makes things easier, but 
I was hoping to avoid any of that. _sigh_ Have I ranted recently about 
how broken the whole packaging & installation thing is?

**Comment by jiffyclub** [comment]
Well, virtualenv makes it easier to install Python packages as non-root, it won't help install compiled libraries as non-root, which is what it looks like @pipitone had to do.

**Comment by pipitone** [comment]
Dreamhost has some instructions/scripts on how to do the python
upgrade/pip/virtualenv installs:
http://wiki.dreamhost.com/Python#Automatic_Installation

Looks easy enough.

**Comment by wking** [comment]
On Mon, Jan 14, 2013 at 09:42:26AM -0800, Greg Wilson wrote:

> As far as I can tell, only Python 2.6 and 2.7 are installed on
> dreamhost right now --- and yeah, we can virtualenv if it makes
> things easier, but I was hoping to avoid any of that. _sigh_ Have I
> ranted recently about how broken the whole packaging & installation
> thing is?

The problem with non-root packaging is that you are relying on a bunch
of software that your personal package manager (pip, homebrew, Gentoo
Prefix, …) doesn't have control over.  If DreamHost doesn't have a
good package manager in place for you to use and you need more than
pure Python packages, it seems like it would be easier to have
deployment folks install the build tools locally, and just rsync the
compiled site to DreamHost.

**Comment by wking** [comment]
On Mon, Jan 14, 2013 at 12:19:37PM -0800, Greg Wilson wrote:

> It would be easy enough to modify the Makefile so that 'make install' 
> rsync'd; only concern there is that sooner or later, someone is going to 
> do it when they're not in the 'master' branch locally (something which 
> doesn't happen now, because we keep 'website' on software-carpentry.org 
> with a single remote, which is website/master).

Minor Git nit: you have a single remote on s-c.o, which is
swcarpentry/website (probably called 'origin').  Unless you've altered
the fetch refspec (see remote.<name>.fetch in `git config --help` and
the refspec notes in `git fetch --help`), this means you've still got
remote tracking branches for anything on swcarpentry/website.  I
expect you meant that you only ever checkout the master branch on
s-c.o.

About automated pushing, it would be easy enough to do:

  BRANCH = $(shell git symbolic-ref --short HEAD)
  publish:
  ifeq ($(BRANCH),master)
    rsync …
  else
    @echo "you're on $(BRANCH), checkout and build master before publishing"
  endif

of course, that doesn't guarantee that they actually rebuild the
output once they checkout `master`, but there's only so much you can
do ;).

**Comment by pipitone** [comment]
bb7e27f merges in Greg's changes to switch to lxml

**Comment by pipitone** [comment]
The switch to lxml seems to have been working for the last week or so with no major issues.  I'm closing this.

**Comment by pipitone** [comment]
I saw that, but when I do a "make check" right now on Amherst, the only
error I get is:

```
make[1]: Entering directory `/home/swcarpentry/website'
    errors in
```

/home/swcarpentry/website/build/bootcamps/2013-02-amos.html
        ((200, 4), 'unexpected-end-tag', {'name': 'p'})

which I just figured had slipped through the cracks.  This isn't a problem
with lxml is it?

**Comment by jiffyclub** [comment]
I think that's a recently added file.

**Comment by pipitone** [comment]
Yeah, exactly.  So... we're good, no?

**Comment by gvwilson** [comment]
That one arises because you can't wrap a <p> around a <table>, but 
without that, the spacing looks weird (not enough space after the table).
G

**Comment by gvwilson** [comment]
If it builds cleanly except for that file, then yes.
G

**Comment by wking** [comment]
On Fri, Jan 25, 2013 at 12:04:11PM -0800, Greg Wilson wrote:

> That one arises because you can't wrap a <p> around a <table>, but 
> without that, the spacing looks weird (not enough space after the table).

That sounds like a job for CSS ;).

**Comment by jiffyclub** [comment]
Or `<br>`.

**Comment by gvwilson** [comment]
CSS please: take out the <p> around the <table> on 
bootcamps/2013-02-amos.html, see what happens to the spacing, work 
magic, send pull request?
Thanks,
Greg

**Comment by jiffyclub** [comment]
I'll fix it. Easiest thing to do is just add a [Bootstrap class](http://twitter.github.com/bootstrap/base-css.html#tables).

**Comment by jiffyclub** [comment]
Fixed in e019544, now looks like this:

![Screen Shot 2013-01-25 at 8 50 12 PM](https://f.cloud.github.com/assets/920492/99355/01818aec-675b-11e2-9be0-6b7223d14533.png)
